openedx/openedx-platform
# [BD-29] [TNL-7262] Course Outline API: Self-paced Course Support

I compared how self-paced course differs from instructor-paced from the backend perspective, and it seems that the only difference is how `hide_after_due` scheduling attribute works for sequences in self-paced course. 


So, when course is instructor-paced, platform respects due date of sequence to decide if block should be hidden:
![Screenshot_2020-07-25_21-51-10](https://user-images.githubusercontent.com/18251194/88464157-3287ec00-cec1-11ea-9f56-79eca4837c3c.png)

But when it's self-paced, sequences due dates don't matter, and some certain block with this attribute enabled, will be hidden only after course end.
![Screenshot_2020-07-25_21-50-40](https://user-images.githubusercontent.com/18251194/88464156-30be2880-cec1-11ea-97c7-bad983a19dae.png)

Here is the code responsible for this behavior:

https://github.com/edx/edx-platform/blob/61e1eda20df2825a409db3e2d36c69d7c36d3e2d/lms/djangoapps/course_blocks/transformers/hidden_content.py#L104-L115

So, to reach the same behavior for the new API, I added new `self_paced` field to `CourseOutlineData` structure and `LearningContext` model. Also, I slightly changed `ScheduleOutlineProcessor`: now it can hide sequences in self-paced courses only for ended courses.

**JIRA tickets**:
- [TNL-7262](https://openedx.atlassian.net/browse/TNL-7262)
- [BLENDED-516](https://openedx.atlassian.net/browse/BLENDED-516)

**Dependencies**: None

**Sandbox URL**: https://pr24609.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

1. Pull changes from this branch to `edx-platform` of your devstack.
1. Apply new migration.
1. Create course (or you can use existing one) with a few sections.
1. Go to `Studio` -> `Settings` -> `Schedule & Details`, change course pacing to `Self-Paced`.
1. Copy course key from address bar.
1. Go to devstack root directory and launch shell:
    ```sh
    make lms-shell
    ```
1. Generate outline of newly created course for new API:
    ```sh
    python manage.py lms update_course_outline <course_key>
    ```
1. Open `http://localhost:18000/api/learning_sequences/v1/course_outline/<course_key>` in browser.
1. You should receive course outline with `self_paced` field and value should be `true`.

**Reviewers**
